### PR TITLE
AES-GCM enabled with AVX512 vAES and vPCLMULQDQ.

### PR DESCRIPTION
**Pull request by shirsing0712**
Performance data: data collected on instance type M6i-2xlarge

awslc_bm speed seal op/sec

Payload in Bytes     1350        8192         16384

AES-128-GCM
  Baseline           2262000      667000       364000
  Patched            2667000     1176000      700000
  Speedup            1.18             1.76              1.92

AES-192-GCM
  Baseline           2173000      619000       337000
  Patched            2585000     1117000      662000
  Speedup            1.19              1.80            1.96

AES-256-GCM
  Baseline           2012000      571000       311000
  Patched            2427000     1053000      624000
  Speedup            1.21             1.84              2.01

Issues:
None

Description of changes: 
Updating AES-GCM implementation to utilize the enhanced crypto ISA's AVX512 vAES and vPCLMULQDQ. Please refer the above details for the performance numbers.

Call-outs:
Nothing known. 

Testing:
ninja-build run_tests

Also, verified the new code path with a sample code for encryption and decryption. Collected the performance number using the benchmark tool in "aws-lc".

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

**Comment by dkostic** [comment]
Hi Shireesh, thanks for opening the PR! There are some peculiarities in our CI that make some of the tests fail. Also, the build logs are not accessible outside of AWS network. But no worries, we'll work with you to fix the issues.

**Comment by dkostic** [comment]
Hi Shireesh, I fixed some of the issues with the code:
- the Perl script input arguments are handled a bit differently in AWS-LC than in OpenSSL, so I fixed that
- We have a folder where we keep generated assembly files for all the platforms we support (this is for customers that can't use Perl/Go in their build environments). So I just run `python util/generate_build_files.py` to generate the new assembly files coming from your Perl scipt.
- We support a build flavor for customers that use really old compilers/assemblers that can't assemble AVX instructions. This is done via build flag. So I added the required handling for that as well.

The remaining issues:
- In the Perl script you generate random suffix for some labels when producing the assembly file. As mentioned above, we add the generated assembly file in the repo. The CI regenerates the assembly file and compares it with the one that's checked in, and fails if they are not the same. Since your Perl script generates different random labels each time it's run, the CI always fails because the labels in the pre-generated and the newly generated file are not the same. To fix this we can either remove the random labels (if they are not necessary) or make the generation pseudo-random (hard-code a seed). What do you think?
- Looks like some old assemblers know about AVX but don't know about AVX512, so I'll have to look into that.
- The Formal Verification tests are failing because they are expecting the old AES-GCM implementation. We'll fix this later, when we are closer to merging the PR.

We are also starting to review the actual code, but it will take some time considering the amount of code :)

**Comment by dkostic** [comment]
Update:
- I made the random label generation in the Perl script deterministic, let's see if the CI is happy with that.
- From our build logs, it seems that gcc-7 on AmazonLinux2 and gcc-4 on CentOS7 are not aware of `aes*` and `vpclmulqdq` instructions. I'm thinking maybe we can substitute the instructions with their bytecode at the end of the Perl script (I think I did this previously for some other code, I'll try to find the example). @shirsing0712 what do you think?

**Comment by dkostic** [comment]
Oh, and just to give a more complete benchmark data from my m6i instance:
```
                Operation                |    patched    |   baseline     |
  EVP-AES-128-GCM Encrypt     (16 bytes) |   4704346.1	 |   4861893.0	  |  0.97
  EVP-AES-128-GCM Decrypt     (16 bytes) |   4882642.6	 |   4813297.3	  |  1.01
  EVP-AES-128-GCM Encrypt    (256 bytes) |   4078340.9	 |   3987586.5	  |  1.02
  EVP-AES-128-GCM Decrypt    (256 bytes) |   4328520.6	 |   3928698.9	  |  1.10
  EVP-AES-128-GCM Encrypt   (1350 bytes) |   2715983.7	 |   2292394.5	  |  1.18
  EVP-AES-128-GCM Decrypt   (1350 bytes) |   2700846.1	 |   2359556.4	  |  1.14
  EVP-AES-128-GCM Encrypt   (8192 bytes) |   1191100.7	 |    679803.5	  |  1.75
  EVP-AES-128-GCM Decrypt   (8192 bytes) |   1202249.8	 |    685041.6	  |  1.76
  EVP-AES-128-GCM Encrypt  (16384 bytes) |    700707.1	 |    367700.7	  |  1.91
  EVP-AES-128-GCM Decrypt  (16384 bytes) |    705060.9	 |    370331.2	  |  1.90
  EVP-AES-192-GCM Encrypt     (16 bytes) |   4680506.6	 |   4806730.8	  |  0.97
  EVP-AES-192-GCM Decrypt     (16 bytes) |   4863172.2	 |   4744569.7	  |  1.02
  EVP-AES-192-GCM Encrypt    (256 bytes) |   4041011.6	 |   3912140.5	  |  1.03
  EVP-AES-192-GCM Decrypt    (256 bytes) |   4307267.4	 |   3851845.9	  |  1.12
  EVP-AES-192-GCM Encrypt   (1350 bytes) |   2653241.2	 |   2203656.2	  |  1.20
  EVP-AES-192-GCM Decrypt   (1350 bytes) |   2651021.8	 |   2264766.7	  |  1.17
  EVP-AES-192-GCM Encrypt   (8192 bytes) |   1131315.6	 |    632555.3	  |  1.79
  EVP-AES-192-GCM Decrypt   (8192 bytes) |   1146644.5	 |    636672.8	  |  1.80
  EVP-AES-192-GCM Encrypt  (16384 bytes) |    661890.1	 |    340302.0	  |  1.95
  EVP-AES-192-GCM Decrypt  (16384 bytes) |    667961.9	 |    342540.3	  |  1.95
  EVP-AES-256-GCM Encrypt     (16 bytes) |   4360655.5	 |   4448514.2	  |  0.98
  EVP-AES-256-GCM Decrypt     (16 bytes) |   4524082.6	 |   4417907.2	  |  1.02
  EVP-AES-256-GCM Encrypt    (256 bytes) |   3786200.8	 |   3653974.4	  |  1.04
  EVP-AES-256-GCM Decrypt    (256 bytes) |   4041289.9	 |   3588325.4	  |  1.13
  EVP-AES-256-GCM Encrypt   (1350 bytes) |   2492820.5	 |   2039554.2	  |  1.22
  EVP-AES-256-GCM Decrypt   (1350 bytes) |   2491387.1	 |   2093344.8	  |  1.19
  EVP-AES-256-GCM Encrypt   (8192 bytes) |   1074842.0	 |    583915.9	  |  1.84
  EVP-AES-256-GCM Decrypt   (8192 bytes) |   1081206.4	 |    587709.1	  |  1.84
  EVP-AES-256-GCM Encrypt  (16384 bytes) |    630155.0	 |    313262.3	  |  2.01
  EVP-AES-256-GCM Decrypt  (16384 bytes) |    628752.9	 |    315892.9	  |  1.99
 AEAD-AES-128-GCM    seal     (16 bytes) |  10859554.5	 |  10813717.6	  |  1.00
 AEAD-AES-128-GCM    seal    (256 bytes) |   7911436.7	 |   7305720.8	  |  1.08
 AEAD-AES-128-GCM    seal   (1350 bytes) |   4000694.0	 |   3118647.6	  |  1.28
 AEAD-AES-128-GCM    seal   (8192 bytes) |   1393107.0	 |    737327.6	  |  1.89
 AEAD-AES-128-GCM    seal  (16384 bytes) |    771988.4	 |    384239.6	  |  2.01
 AEAD-AES-256-GCM    seal     (16 bytes) |  10330064.1	 |  10192387.9	  |  1.01
 AEAD-AES-256-GCM    seal    (256 bytes) |   7489985.0	 |   6831952.2	  |  1.10
 AEAD-AES-256-GCM    seal   (1350 bytes) |   3660191.4	 |   2764809.2	  |  1.32
 AEAD-AES-256-GCM    seal   (8192 bytes) |   1244452.4	 |    630821.5	  |  1.97
 AEAD-AES-256-GCM    seal  (16384 bytes) |    686051.9	 |    327090.4	  |  2.10
```

File: crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl [hunk 1]
```diff
@@ -0,0 +1,4869 @@
...
+  192 => 11,
+  256 => 13);
+
+# ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+# ;;; Code generation control switches
+# ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+# ; ABI-aware zeroing of volatile registers in EPILOG().
+# ; Disabled due to performance reasons.
+my $CLEAR_SCRATCH_REGISTERS = 0;
```
Comment (dkostic) [inline crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl:67]: are we ever going to use this? If not, maybe we should remove the code since the file is already very large.

File: crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl [hunk 2]
```diff
@@ -0,0 +1,4869 @@
...
+
+# ; Zero HKeys storage from the stack if they are stored there
+my $CLEAR_HKEYS_STORAGE_ON_EXIT = 1;
+
+# ; Enable / disable check of function arguments for null pointer
+# ; Currently disabled, as this check is handled outside.
+my $CHECK_FUNCTION_ARGUMENTS = 0;
```
Comment (dkostic) [inline crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl:74]: are we ever going to use this? If not, maybe we should remove the code since the file is already very large.

File: crypto/fipsmodule/modes/gcm_test.cc [hunk 3]
```diff
@@ -189,6 +189,40 @@ TEST(GCMTest, ABI) {
...
+      CHECK_ABI_SEH(gcm_gmult_avx512, X, Htable);
+      for (size_t blocks : kBlockCounts) {
+        CHECK_ABI_SEH(gcm_ghash_avx512, X, Htable, buf, 16 * blocks);
+      }
+
+      if (hwaes_capable()) {
+        AES_KEY aes_key;
+        static const uint8_t kKey[16] = {0};
+
+        // aes_gcm_*_avx512 makes assumptions about |GCM128_CONTEXT|'s layout.
```
Comment (dkostic) [inline crypto/fipsmodule/modes/gcm_test.cc:203]: can we add a unit test to make sure the assumption is correct?

**Comment by dkostic** [comment]
From the build logs, looks like the only remaining issue is with old compilers not being aware of `aes*` and `vpclmulqdq` instructions. It'd be great if we could work around this by adding the instructions bytecode. Otherwise, I'll have to disable compiling the new implementation when too old compilers are used.

File: crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl (hunk 1 above)
Comment (shirsing0712) [reply crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl:67]: will do the cleanup.

File: crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl (hunk 2 above)
Comment (shirsing0712) [reply crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl:74]: same

File: crypto/fipsmodule/modes/gcm_test.cc (hunk 3 above)
Comment (shirsing0712) [reply crypto/fipsmodule/modes/gcm_test.cc:203]: This is inherited from the old code (line 168), not sure how can we add  UT for structure layout.

**Comment by shirsing0712** [comment]
> 



> Update:
> 
> * I made the random label generation in the Perl script deterministic, let's see if the CI is happy with that.
> * From our build logs, it seems that gcc-7 on AmazonLinux2 and gcc-4 on CentOS7 are not aware of `aes*` and `vpclmulqdq` instructions. I'm thinking maybe we can substitute the instructions with their bytecode at the end of the Perl script (I think I did this previously for some other code, I'll try to find the example). @shirsing0712 what do you think?

Yes, we can add the bytecode, could you please help with example?

**Comment by dkostic** [comment]
> Yes, we can add the bytecode, could you please help with example?

Here's how it's done for AES-NI instructions for example, but we also need it for vpclmul:
- https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/aes/asm/aesni-x86_64.pl#L5057-L5102

**Comment by shirsing0712** [comment]
> > 
> 
> > Update:
> > 
> > * I made the random label generation in the Perl script deterministic, let's see if the CI is happy with that.
> > * From our build logs, it seems that gcc-7 on AmazonLinux2 and gcc-4 on CentOS7 are not aware of `aes*` and `vpclmulqdq` instructions. I'm thinking maybe we can substitute the instructions with their bytecode at the end of the Perl script (I think I did this previously for some other code, I'll try to find the example). @shirsing0712 what do you think?
> 
> Yes, we can add the bytecode, could you please help with example?

@dkostic for now, we wanted to disable the AVX512 support for old assemblers. Shall we introduce another flag for that (something like MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX512)?

**Review by dkostic — APPROVED** [review]

**Comment by shirsing0712** [comment]
@nebeid Please let me know if you have any feedback?

**Comment by nebeid** [comment]
> @nebeid Please let me know if you have any feedback?

Thank you for asking, @shirsing0712.
I have the following recommendations:
- Can we rename the Perl file to start with the prefix aesni to be consistent with the other Intel implementation aesni-gcm-x86_64.pl in the same folder?
- Is it possible to put a general description at the top of the file about the main loop operation. Something similar to https://github.com/aws/aws-lc/blob/990f98ce50b9cb05b295161007190da4fe3bacbe/crypto/fipsmodule/modes/asm/aesv8-gcm-armv8.pl#L26-L27
  I see you have a description starting at https://github.com/aws/aws-lc/blob/990f98ce50b9cb05b295161007190da4fe3bacbe/crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl#L3613 of what I think is the general description of the main loop. It would be good if it is put as a paragraph about the main optimisation idea at the top of the file.
- I think we may need to add test cases for the specific branching edge cases in the code. They may be many, so you can choose based on the different branches which ones are to be tested.

File: crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl [hunk 4]
```diff
@@ -0,0 +1,4822 @@
...
+  $arg9  = "`$GP_STORAGE + 8*3`(%rbp)";
+  $arg10 = "`$GP_STORAGE + 8*4`(%rbp)";
+  $arg11 = "`$GP_STORAGE + 8*5`(%rbp)";
+}
+
+# ; Offsets in gcm128_context structure (see crypto/fipsmodule/modes/modes.h)
+my $CTX_OFFSET_CurCount  = (16 * 0);          #  ; (Yi) Current counter for generation of encryption key
+my $CTX_OFFSET_PEncBlock = (16 * 1);          #  ; (repurposed EKi field) Partial block buffer
+my $CTX_OFFSET_EK0       = (16 * 2);          #  ; (EK0) Encrypted Y0 counter (see gcm spec notation)
+my $CTX_OFFSET_AadLen    = (16 * 3);          #  ; (len.u[0]) Length of Hash which has been input
```
Comment (nebeid) [inline crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl:147]:
```suggestion
my $CTX_OFFSET_AadLen    = (16 * 3);          #  ; (len.u[0]) Length of AAD which has been input
```

**Comment by nebeid** [comment]
>     * I think we may need to add test cases for the specific branching edge cases in the code. They may be many, so you can choose based on the different branches which ones are to be tested.

For example, I added some test cases for the 8x-unrolling Arm code in PR https://github.com/aws/aws-lc/pull/721 under cipher_extra/test.

**Comment by shirsing0712** [comment]
> > ```
> > * I think we may need to add test cases for the specific branching edge cases in the code. They may be many, so you can choose based on the different branches which ones are to be tested.
> > ```
> 
> For example, I added some test cases for the 8x-unrolling Arm code in PR #721 under cipher_extra/test.

@nebeid thank you for the feedback, we are looking into it.

**Comment by tkanteck** [comment]
>   I see you have a description starting at https://github.com/aws/aws-lc/blob/990f98ce50b9cb05b295161007190da4fe3bacbe/crypto/fipsmodule/modes/asm/aes-gcm-avx512.pl#L3613
>    of what I think is the general description of the main loop. It would be good if it is put as a paragraph about the main optimisation idea at the top of the file.

This is good suggestion.

The main building block of the loop is code that finely stitches AES-CTR and GHASH functions on 16 blocks of data. It uses VAES and VPCLMULQDQ instructions with full width of ZMM registers. AES-CTR and GHASH execute in parallel to large extend.
The main loop executes this 16 block parallel code three times. Consequently, it processes 3 x 16 = 48 blocks of data and GHASH reduction is done only at the end of the loop (once per 48 blocks).
In the main loop, AES-CTR cipher is ahead of GHASH by 32 blocks. For example, when in 16 block parallel code AES-CTR processes blocks number 48 to 63 then GHASH processes blocks number 16 to 31.

@shirsing0712 could you add this paragraph to the code, please?

**Comment by shirsing0712** [comment]
@nebeid I have updated the patch, could you please take a look?

**Review by nebeid — DISMISSED** [review]

**Review by dkostic — APPROVED** [review]

**Review by nebeid — APPROVED** [review]
